File: modules/jiosaavn.py
import requests

def search_jiosaavn(query):
    # JioSaavn Hidden API
    url = "https://www.jiosaavn.com/api.php"
    params = {
        "__call": "search.getResults",
        "_format": "json",
        "_marker": "0",
        "api_version": "4",
        "p": "1",
        "n": "10",
        "q": query
    }
    
    results = []
    try:
        res = requests.get(url, params=params)
        data = res.json()
        
        if "results" in data:
            for song in data["results"]:
                more_info = song.get("more_info", {})
                
                # Get Title
                title = song.get("title")
                if not title: continue # Skip invalid results

                # Get Artist
                artist = "Unknown"
                if "artistMap" in more_info:
                    p_artists = more_info["artistMap"].get("primary_artists", [])
                    if p_artists:
                        artist = p_artists[0].get("name")
                
                # Fallback Artist
                if artist == "Unknown":
                    artist = song.get("subtitle", "Unknown")

                # Get Duration
                duration = 0
                try:
                    duration = int(more_info.get("duration", 0))
                except:
                    pass
                
                # High-res image fix
                img = song.get("image", "").replace("150x150", "500x500")
                
                results.append({
                    "title": title,
                    "artist": artist,
                    "duration": duration,
                    "id": song.get("id"),
                    "url": song.get("perma_url"), 
                    "thumbnail": img,
                    "source": "jiosaavn"
                })
    except Exception as e:
        print(f"JioSaavn Error: {e}")
        
    return results

File: backend/plugins/jiosaavn.js
const axios = require("axios");

const BASE = "https://www.jiosaavn.com/api.php";

async function searchSongs(query) {
  try {
    const res = await axios.get(BASE, {
      params: {
        __call: "search.getResults",
        _format: "json",
        _marker: "0",
        api_version: "4",
        ctx: "web6dot0",
        query,
        n: "10",
        p: "1"
      },
      timeout: 8000
    });

    const results = res.data?.results || [];
    return results.map((item) => ({
      id: item.id,
      title: item.title?.replace(/&amp;/g, "&") || "Unknown",
      artist: item.more_info?.singers || item.subtitle || "Unknown",
      thumbnail: item.image?.replace("150x150", "500x500") || "",
      duration: parseInt(item.more_info?.duration || "0"),
      source: "jiosaavn"
    }));
  } catch {
    return [];
  }
}

async function getStreamUrl(id) {
  try {
    const res = await axios.get(BASE, {
      params: {
        __call: "song.generateAuthToken",
        _format: "json",
        _marker: "0",
        api_version: "4",
        ctx: "web6dot0",
        bitrate: "320",
        url: `https://www.jiosaavn.com/song/x/${id}`
      },
      timeout: 8000
    });

    const url = res.data?.auth_url;
    return { url: url || null };
  } catch {
    return { url: null };
  }
}

module.exports = { name: "jiosaavn", searchSongs, getStreamUrl };
